pkg/virt-controller/watch/application.go

```diff
@@ -105,14 +109,63 @@ func (vca *VirtControllerApp) Run() {
 	stop := make(chan struct{})
 	defer close(stop)
 	vca.informerFactory.Start(stop)
-	go vca.vmController.Run(3, stop)
-	go vca.migrationController.Run(3, stop)
-	go vca.rsController.Run(3, stop)
-	httpLogger := logger.With("service", "http")
-	httpLogger.Info().Log("action", "listening", "interface", vca.host, "port", vca.port)
-	if err := http.ListenAndServe(vca.host+":"+strconv.Itoa(vca.port), nil); err != nil {
+	go func() {
+		httpLogger := logger.With("service", "http")
+		httpLogger.Info().Log("action", "listening", "interface", vca.host, "port", vca.port)
+		if err := http.ListenAndServe(vca.host+":"+strconv.Itoa(vca.port), nil); err != nil {
+			golog.Fatal(err)
+		}
+	}()
+
+	recorder := vca.getNewRecorder(k8sv1.NamespaceAll, leaderelectionconfig.DefaultEndpointName)
+
+	id, err := os.Hostname()
+	if err != nil {
+		golog.Fatalf("unable to get hostname: %v", err)
+	}
+
+	rl, err := resourcelock.New(vca.LeaderElection.ResourceLock,
+		leaderelectionconfig.DefaultNamespace,
+		leaderelectionconfig.DefaultEndpointName,
+		vca.clientSet.CoreV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      id,
+			EventRecorder: recorder,
+		})
+	if err != nil {
 		golog.Fatal(err)
 	}
+
+	leaderElector, err := leaderelection.NewLeaderElector(
+		leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: vca.LeaderElection.LeaseDuration.Duration,
+			RenewDeadline: vca.LeaderElection.RenewDeadline.Duration,
+			RetryPeriod:   vca.LeaderElection.RetryPeriod.Duration,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(stopCh <-chan struct{}) {
```

Comment:
Is it ok, if this function just starts go-routines and continues? Should it block?

Reply:
If I understand correct `leaderElector.Run()` just run callback `OnStartedLeading` function and enter to infinitive loop that tries to renew a lease, so it must be ok.

Reply:
I took a look at kubernete's controllers. They end the OnStartedLeading function with a select{} statement, which blocks indefinitely.  I guess we should do the same

Reply:
It really shouldn't matter though. I think the select{} isn't there so much to ensure that callback blocks as it is to ensure a deadlock doesn't occur which leaves all goroutines idle.

Reply:
ok, so not needed. Thx for investigating.